src/charm.py

```diff
@@ -45,21 +53,68 @@ def __init__(self, *args):
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(self.on[PEER].relation_changed, self._on_peer_relation_changed)
+        self.framework.observe(self.on[PEER].relation_departed, self._on_peer_relation_departed)
         self.framework.observe(self.on.postgresql_pebble_ready, self._on_postgresql_pebble_ready)
+        self.framework.observe(self.on.stop, self._on_stop)
         self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
         self.framework.observe(
             self.on.get_postgres_password_action, self._on_get_postgres_password
         )
         self.framework.observe(self.on.get_primary_action, self._on_get_primary)
         self.framework.observe(self.on.update_status, self._on_update_status)
         self._storage_path = self.meta.storages["pgdata"].location
-        self._patroni = Patroni(self._pod_ip, self._storage_path)
+
+    @property
+    def _endpoints_to_remove(self):
+        """List the endpoints that were part of the cluster but departed."""
+        old = self._endpoints
+        current = [self._get_hostname_from_unit(member) for member in self._hosts]
+        endpoints_to_remove = list(set(old) - set(current))
+        return endpoints_to_remove
+
+    def _on_peer_relation_departed(self, _):
+        """The leader removes the departing units from the list of cluster members."""
+        if not self.unit.is_leader():
+            return
+
+        self._update_endpoints(endpoints_to_remove=self._endpoints_to_remove)
+
+    def _on_peer_relation_changed(self, event: RelationChangedEvent):
+        """Reconfigure cluster members."""
+        # The cluster must be initialized first in the leader unit
+        # before any other member joins the cluster.
+        if "cluster_initialised" not in self._peers.data[self.app]:
+            event.defer()
+            return
+
+        # If the leader is the one receiving the event, it adds the new members,
+        # one at a time.
+        if self.unit.is_leader():
+            self._reconfigure(event)
+            return
+
+        if self._endpoint in self._endpoints:
+            # Update the list of the cluster members in the replicas to make them know each other.
+            try:
+                # Update the cluster members in this unit (updating patroni configuration).
+                self._patroni.update_cluster_members()
+                # Add the labels needed for replication in this pod.
+                self._patch_pod_labels(self._unit)
+                # Validate the status of the member before setting an ActiveStatus.
+                try:
+                    if not self._patroni.member_started:
+                        raise NotReadyError
+                except (NotReadyError, RetryError):
```

Comment:
would it be possible to avoid having nested try/except blocks? these are usually really hard to trace and debug if need be in the future

Reply:
Of course. I mixed some try/except blocks here. Thanks for catching this! Fixed on https://github.com/canonical/postgresql-k8s-operator/pull/9/commits/eaf5d47b081a0489886e949ee5d4938088cdda6d.